README.md
# sinopia cookbook
[![CK Version](http://img.shields.io/cookbook/v/sinopia.svg)](https://supermarket.getchef.com/cookbooks/sinopia) [![Build Status](https://img.shields.io/travis/redguide/nodejs.svg)](https://travis-ci.org/BarthV/sinopia-cookbook)

Sinopia project : https://github.com/rlidwka/sinopia/

Sinopia is a private/caching npm repository server.

It allows you to have a local npm registry with zero configuration. You don't have to install and replicate an entire CouchDB database. Sinopia keeps its own small database and, if a package doesn't exist there, it asks npmjs.org for it keeping only those packages you use.

# Requirements

- Chef >= 11
- Supported platforms : Ubuntu 12.04, 14.04 & CentOS 6, 7.
- Sinopia Cookbook heavily depends on `nodejs` community cookbooks.

# Usage

Add `recipe[sinopia]` to the node runlist.

Default recipe with no other options will :

- Configure sinopia folders (in /etc, /var & /var/log)
- Install node + npm from the official repo at the latest version
- Create a passwordless sinopia user who will run the service
- Install latest version of sinopia from npmjs.org
- Configure log rotation to 30d
- Setup and start sinopia service (based on upstart for now)

## Defaults
* Sinopia will bind to `127.0.0.1:4873`, so you probably need to setup a web frontend.
* Access to the npm service is allowed to everyone.
* All desired packages are cached from https://registry.npmjs.org/
* A single npm account is provisionned to publish private packages with :
 * login : `admin`
 * passw : `admin`

# Attributes
Every single Sinopia configuration item can be managed from node attributes.
Default values are specified each time.

## System configuration

- `node['sinopia']['user']` : (sinopia) default user running sinopia
- `node['sinopia']['confdir']` : (/etc/sinopia) config.yaml file location
- `node['sinopia']['datadir']` : (/var/lib/sinopia) Sinopia cache & private stores location
- `node['sinopia']['logdir']` : (/var/log/sinopia) sinopia.log file location
- `node['sinopia']['logdays']` : (30) log retention policy, `Integer` days
- `node['sinopia']['loglevel']` : (warn) log level, trace | debug | info | http | warn | error | fatal

## Sinopia global conf

- `node['sinopia']['version']` : (nil) Sinopia npm package version, use `nil` for latest
- `node['sinopia']['admin']['pass']` : (admin) Sinopia admin account clear password
- `node['sinopia']['public_url']` : Sinopia rewrite url, url prefix for provided links
- `node['sinopia']['timeout']` : (nil) Cached repo timeout in ms, software default is 30000 ms
- `node['sinopia']['maxage']` : (nil) Sinopia metadata cache max age in sec, software defaut is 120s
- `node['sinopia']['max_body_size']` : (nil) Maximum size of uploaded json document, software default is 1mb

## Users and rights

No users are created by default.

* You can set user list with a hash under `default['sinopia']['users']`, you need to specify a password for each user
* You can give admin right to a specific user with `user['admin'] = true` hash

Example:
```ruby
node['sinopia']['users']['bob']['pass'] = 'incredible'
node['sinopia']['users']['bob']['admin'] = true

node['sinopia']['users']['andy']['pass'] = 'toys'
node['sinopia']['users']['andy']['admin'] = true

node['sinopia']['users']['woody']['pass'] = 'buzz'
```

## NPM Registry

You can store a list of available npm repositories in `node['sinopia']['repos']` following {'name' => 'url'} syntax. 

Default hash is loaded with official npmjs repo : `default['sinopia']['repos'] = {'npmjs' => 'https://registry.npmjs.org/'}`

Example :
```ruby
node['sinopia']['repos'] = {
  'npmjs' => 'https://registry.npmjs.org/', # official npmjs repo
  'myrepo' => 'https://myrepo.local/',
  'other' => 'https://third-party-repo.com'
}
```

`node['sinopia']['mainrepo']` : (npmjs) Caching repository name selected from available repos list

## Filters

- `default['sinopia']['strict_access']` : When set to `true`, this only allow admin and admin users to access sinopia repos, default is `false`
- You can define access & publish filters based on package name under `default['sinopia']['filters']`
- Filter format is an Array with one Hash for one rule  
- Wildcard is accepted in the filter name rule
- Access can be provided to :
 * Default (all)
 * Specified available users : `['user1', 'user2']`
 * admin account + all admin user : '@admins'
- publish can be provided to :
 * Default (admin account only)
 * Specified available users + admin : `['user1', 'user2']`
 * admin account + all admin user : '@admins'
- Storage value is the name of the folder where filtered packages will be set.

Example :
```ruby
node['sinopia']['filters'] = [
  {
    'name' => 'private-*',
    'storage' => 'private-repo'
  },
  {
    'name' => 'admin-*',
    'access' => ['andy', 'woody']
  },
  {
    'name' => 'test-*',
    'access' => '@admins'
  }
]
```

## Logging

This cookbook is reusing specific logging format of Sinopia :

```
type: file | stdout | stderr
level: trace | debug | info | http (default) | warn | error | fatal

{type: 'file', path: 'sinopia.log', level: 'debug'},

parameters for stdout and stderr: format: json | pretty
{type: 'stdout', format: 'pretty', level: 'debug'}
```

You can add as much logger as you want (including '{}') in `default['sinopia']['logs']` Array

Default value is :
```ruby
node['sinopia']['logs'] = [
  "{type: file, path: '/var/log/sinopia/sinopia.log', level: warn}"
]
```

## Proxy

See `attributes/default.rb` to view how to configure `node['sinopia']['use_proxy']` and `node['sinopia']['proxy']`.

## NPM

See `attributes/default.rb` to view Node & npm install options (version, source/package, ...)

# Recipes

`sinopia::default` recipe includes :
- `sinopia::users` : creates users
- `sinopia::nodejs` : install node & npm
- `sinopia::sinopia` : install sinopia, directories, conf and start service

# Testing

Sinopia cookbook is bundled with a Vagrantfile. If you have virtualbox and vagrant ready, just fire a `vagrant up` and this will setup a box running Sinopia and listening 0.0.0.0:4873. Port 4873 is forwaded to your 127.0.0.1:4873 for test purposes.

# Author

Author:: Barthelemy Vessemont (<bvessemont@gmail.com>)


CHANGELOG.md
Sinopia Cookbook CHANGELOG
==========================

0.3.2
-------------------
- Adding changelog file
- Adding EL6/EL7 support
- Adding systemd support for debian platform
- fixing ACL in config file


spec/npmlazy_default_spec.rb
# require 'spec_helper'
#
# describe 'npmlazy::nodejs' do
#  context 'on Ubuntu 12.04' do
#    let(:chef_run) do
#      ChefSpec::Runner.new(platform: 'ubuntu', version: '12.04')
#      .converge(described_recipe)
#    end
#
#    it 'installs required packages' do
#      %w{ make }.each do |package|
#        expect(chef_run).to install_package(package)
#      end
#    end
#  end
# end

#
#   context 'on Centos 5.9' do
#     let(:chef_run) do
#       ChefSpec::Runner.new(platform: 'centos', version: '5.9').
#         converge(described_recipe)
#     end
#
#     it 'installs required packages' do
#       %w{ php53-mysql php53-gd php53-bcmath php53-mbstring }.each do |package|
#         expect(chef_run).to install_package(package)
#       end
#     end
#   end
#
#   context 'on Centos 6.0' do
#     let(:chef_run) do
#       ChefSpec::Runner.new(platform: 'centos', version: '6.0').
#         converge(described_recipe)
#     end
#
#     it 'installs required packages' do
#       %w{ php php-mysql php-gd php-bcmath php-mbstring php-xml }.each do |package|
#         expect(chef_run).to install_package(package)
#       end
#     end
#   end
# end


spec/sinopia_default_spec.rb
# require 'spec_helper'

# describe 'sinopia::nodejs' do
#  context 'on Ubuntu 12.04' do
#    let(:chef_run) do
#      ChefSpec::Runner.new(platform: 'ubuntu', version: '12.04')
#      .converge(described_recipe)
#    end
#
#    it 'installs required packages' do
#      %w{ make }.each do |package|
#        expect(chef_run).to install_package(package)
#      end
#    end
#  end
# end

#
#   context 'on Centos 5.9' do
#     let(:chef_run) do
#       ChefSpec::Runner.new(platform: 'centos', version: '5.9').
#         converge(described_recipe)
#     end
#
#     it 'installs required packages' do
#       %w{ php53-mysql php53-gd php53-bcmath php53-mbstring }.each do |package|
#         expect(chef_run).to install_package(package)
#       end
#     end
#   end
#
#   context 'on Centos 6.0' do
#     let(:chef_run) do
#       ChefSpec::Runner.new(platform: 'centos', version: '6.0').
#         converge(described_recipe)
#     end
#
#     it 'installs required packages' do
#       %w{ php php-mysql php-gd php-bcmath php-mbstring php-xml }.each do |package|
#         expect(chef_run).to install_package(package)
#       end
#     end
#   end
# end


recipes/sinopia.rb
# encoding: utf-8
#
# Cookbook Name:: sinopia
# Recipe:: sinopia
#

require 'digest'

nodejs_npm 'sinopia' do
  version node['sinopia']['version']
  action :install
end

directory node['sinopia']['confdir'] do
  recursive true
end

directory node['sinopia']['logdir'] do
  owner node['sinopia']['user']
  group node['sinopia']['user']
end

directory node['sinopia']['datadir'] do
  owner node['sinopia']['user']
  group node['sinopia']['user']
end

admin_add_list = []

node['sinopia']['users'].each do |user, conf|
  admin_add_list.push(user) if conf['admin']
end

template File.join(node['sinopia']['confdir'], 'config.yaml') do
  source 'config.yaml.erb'
  variables(
    :admins => admin_add_list
  )
  mode 0444
  notifies :restart, 'service[sinopia]', :delayed
end

logrotate_app 'sinopia' do
  cookbook 'logrotate'
  path File.join(node['sinopia']['logdir'], 'sinopia.log')
  frequency 'daily'
  rotate node['sinopia']['logdays']
  create '644 root adm'
end

case node['platform_family']
when 'rhel'
  if node['init_package'] == 'systemd'
    template '/usr/lib/systemd/system/sinopia.service' do
      source 'sinopia.service.erb'
      mode '0644'
      notifies :run, 'execute[systemd-reload]', :immediately
      notifies :restart, 'service[sinopia]', :delayed
    end
    execute 'systemd-reload' do
      command 'systemctl daemon-reload'
      action :nothing
    end
  else
    package 'redhat-lsb-core'

    template '/etc/init.d/sinopia' do
      source 'sinopia.sysvinit.erb'
      mode '0755'
    end
  end
when 'debian'
  if node['init_package'] == 'systemd'
    template '/lib/systemd/system/sinopia.service' do
      source 'sinopia.service.erb'
      mode '0644'
      notifies :restart, 'service[sinopia]', :delayed
    end
  else
    template '/etc/init/sinopia.conf' do
      source 'sinopia.conf.erb'
      notifies :restart, 'service[sinopia]', :delayed
    end
  end
else
  raise 'Platform not supported'
end

service 'sinopia' do
  supports :status => true, :restart => true, :reload => false
  action [:enable, :start]
end


attributes/default.rb
# encoding: utf-8

## System user running sinopia
default['sinopia']['user'] = 'sinopia'

## sinopia gem version (use nil for latest)
default['sinopia']['version'] = nil

## sinopia users configuration
default['sinopia']['admin']['pass'] = 'admin'

default['sinopia']['users'] = {}

# default['sinopia']['users']['bob']['pass'] = 'incredible'
# default['sinopia']['users']['bob']['admin'] = true

# default['sinopia']['users']['andy']['pass'] = 'toys'
# default['sinopia']['users']['andy']['admin'] = true

# default['sinopia']['users']['woody']['pass'] = 'buzz'

## sinopia links rewrite URL (url_prefix)
# default['sinopia']['public_url'] = 'https://my-npm-private-repo.local/'

## Bind address (IP:port format)
# use nil for default (127.0.0.1:4873)
# use ':port' or '0.0.0.0:port' to listen on all interfaces
default['sinopia']['listen'] = nil

## sinopia conf directories
# Parents directory MUST exists !
default['sinopia']['confdir'] = '/etc/sinopia'
default['sinopia']['logdir'] = '/var/log/sinopia'
default['sinopia']['logdays'] = 30
default['sinopia']['datadir'] = '/var/lib/sinopia'
default['sinopia']['loglevel'] = 'warn'

## NodeJS repo list options
default['sinopia']['repos'] = {
  'npmjs' => 'https://registry.npmjs.org/' # official npmjs repo
  # 'myrepo' => 'https://myrepo.local/'
}

default['sinopia']['mainrepo'] = 'npmjs'

default['sinopia']['timeout'] = nil # 30000 ms
default['sinopia']['maxage'] = nil # 120 s
default['sinopia']['max_body_size'] = nil # 1mb

# Restric read access for admins only
default['sinopia']['strict_access'] = false

default['sinopia']['use_proxy'] = false
default['sinopia']['proxy']['http'] = 'http://something.local/'
default['sinopia']['proxy']['https'] = 'https://something.local/'
default['sinopia']['proxy']['no_proxy'] = [
  'localhost', '127.0.0.1'
]

default['sinopia']['htpasswd']['enable'] = false
default['sinopia']['htpasswd']['path'] = "#{default['sinopia']['datadir']}/.htpasswd"

## local repos ACL - filters
default['sinopia']['filters'] = [
  # {
  #   'name' => 'private-*',
  #   'storage' => 'private-repo'
  # },
  # {
  #   'name' => 'admin-*',
  #   'access' => ['andy', 'woody']
  # },
  #
  ## @admin is a special value for admin account + all admin users
  #
  # {
  #   'name' => 'test-*',
  #   'access' => '@admins'
  # }
]

## Logging options
# type: file | stdout | stderr
# level: trace | debug | info | http (default) | warn | error | fatal
#
# parameters for file: name is filename
#  {type: 'file', path: 'sinopia.log', level: 'debug'},
#
# parameters for stdout and stderr: format: json | pretty
#  {type: 'stdout', format: 'pretty', level: 'debug'},
default['sinopia']['logs'] = [
  "{type: 'file', path: '#{File.join(node['sinopia']['logdir'], 'sinopia.log')}', level: '#{node['sinopia']['loglevel']}'}"
]
